rules-tests/TypeDeclaration/Rector/ClassMethod/AddMethodCallBasedStrictParamTypeRector/Fixture/skip_native_optional_array_shape.php.inc

```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace Rector\Tests\TypeDeclaration\Rector\ClassMethod\AddMethodCallBasedStrictParamTypeRector\Fixture;
+
+final class SkipNativeOptionalShape {
+    private function doFoo() {
+        $shape = pathinfo('');
```

Comment:
it seems `pathinfo()` getNativeType() PHPStan bug, getting native type of `$shape` itself got union array and string:

```php
PHPStan\Type\UnionType #11106
   types: array (2)
   |  0 => PHPStan\Type\ArrayType #11170 ...
   |  1 => PHPStan\Type\StringType #11098 ...
   normalized: true
   sortedTypes: true
   cachedDescriptions: array (1)
   |  4 => 'array|string'
```

Reply:
on native return, `pathinfo()` got `array` or `string` https://www.php.net/manual/en/function.pathinfo.php

Reply:
I think the best way to verify if the shape is array or string, eg:

```diff
        $shape = pathinfo('');
+       assert(is_array($shape));
        $this->doBar($shape['dirname']); // dirname is only conditionally returned
```

then, it will be skipped

Reply:
it seems can be fixed with the following check:

```php
            $nativeVariableType = $scope->getNativeType($expr->var);
            if (! $nativeVariableType instanceof MixedType && (! $nativeVariableType instanceof ArrayType || ! $nativeVariableType->getItemType() instanceof MixedType)) {
                $type = $scope->getType($expr);

                if ($expr->dim instanceof String_) {
                    $targetKey = null;
                    $variableType = $scope->getType($expr->var);
                    if ($variableType instanceof ArrayType) {
                        foreach ($variableType->getKeyTypes() as $key => $type) {
                            if ($type instanceof \PHPStan\Type\Constant\ConstantStringType && $type->getValue() === $expr->dim->value) {
                                $targetKey = $key;
                                break;
                            }
                        }

                        if ($targetKey !== null && in_array($targetKey, $variableType->getOptionalKeys(), true)) {
                            $type = $scope->getNativeType($expr);
                        }
                    }
                }
            }
```

Reply:
https://github.com/rectorphp/rector-src/pull/5062